relay-general/src/pii/attachments.rs

```diff
@@ -269,75 +267,58 @@ impl StringMods for [u8] {
 /// encoded text.
 struct WStrSegmentIter<'a> {
     data: &'a mut [u8],
-    decoder: Box<dyn RawDecoder>,
     offset: usize,
 }
 
 impl<'a> WStrSegmentIter<'a> {
     fn new(data: &'a mut [u8]) -> Self {
-        Self {
-            data,
-            decoder: UTF_16LE.raw_decoder(),
-            offset: 0,
-        }
+        Self { data, offset: 0 }
     }
 }
 
 impl<'a> Iterator for WStrSegmentIter<'a> {
-    type Item = WstrSegment<'a>;
+    type Item = WStrSegment<'a>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        // We are handing out multiple mutable slices from the same mutable slice.  This is
-        // safe because we know they are not overlapping.  However the compiler doesn't know
-        // this so we need to transmute the lifetimes of the slices we return with
-        // std::slice::from_raw_parts_mut().
-        let mut decoded = String::with_capacity(self.data.len() - self.offset);
-
         loop {
             if self.offset >= self.data.len() {
                 return None;
             }
 
-            decoded.clear();
-            let start = self.offset;
-            let (unprocessed_offset, err) =
-                self.decoder.raw_feed(&self.data[start..], &mut decoded);
-            let end = start + unprocessed_offset;
-
-            if let Some(err) = err {
-                if err.upto > 0 {
-                    self.offset += err.upto as usize;
-                } else {
-                    // This should never happen, but if it does, re-set the decoder and skip
-                    // forward to the next 2 bytes.
-                    self.offset += std::mem::size_of::<u16>(); // TODO: encoding-neutral?!?
-                    self.decoder = self.decoder.from_self();
-                }
-                if decoded.len() > 2 {
-                    let slice = &mut self.data[start..end];
-                    let len = slice.len();
-                    let ptr = slice.as_mut_ptr();
-                    let encoded = unsafe {
-                        WStr::from_utf16le_unchecked_mut(std::slice::from_raw_parts_mut(ptr, len))
-                    };
-                    return Some(WstrSegment { encoded, decoded });
-                } else {
-                    continue;
+            let slice: &mut [u8] = match WStr::from_utf16le_mut(&mut self.data[self.offset..]) {
```

Comment:
The compiler should be able to infer this.

```suggestion
            let slice = match WStr::from_utf16le_mut(&mut self.data[self.offset..]) {
```

Reply:
uhm, yes.  this was helpful while i was getting the returned types right but forgot to clean up